### PR TITLE
put count back

### PR DESCRIPTION
**Pull request by dogfish182**
As this module is nested within the VPC module, trying to use for_each produces the following error in terraform. 

```
Error: Invalid for_each argument

  on .terraform/modules/network.subnet_sharing/main.tf line 9, in resource "aws_ram_resource_association" "private_subnets":
   9:   for_each           = toset(var.private_subnet_arns)

The "for_each" value depends on resource attributes that cannot be determined
until apply, so Terraform cannot predict how many instances will be created.
To work around this, use the -target argument to first apply only the
resources that the for_each depends on.


Error: Invalid for_each argument

  on .terraform/modules/network.subnet_sharing/main.tf line 15, in resource "aws_ram_resource_association" "public_subnets":
  15:   for_each           = toset(var.public_subnet_arns)

The "for_each" value depends on resource attributes that cannot be determined
until apply, so Terraform cannot predict how many instances will be created.
To work around this, use the -target argument to first apply only the
resources that the for_each depends on.

```

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -1,18 +1,16 @@
 resource "aws_ram_resource_share" "subnet_sharing" {
-  count                     = length(var.private_subnet_arns) > 0 || length(var.public_subnet_arns) > 0 ? 1 : 0
+  count                     = length(var.private_subnet_arns) >= 0 || length(var.public_subnet_arns) >= 0 ? 1 : 0
```
Comment (svanharmelen) [inline main.tf]: This is a bug which we already discussed last week.. `>=`

File: main.tf [hunk 2]
```diff
@@ -1,18 +1,16 @@
...
   name                      = "${var.prepend_resource_type ? "resource-share-" : ""}subnets-${var.name}"
   allow_external_principals = true
-  tags                      = var.tags
+  tags                      = merge({ resource-type = "ec2:Subnet" }, var.tags)
```
Comment (svanharmelen) [inline main.tf]: We agreed not to do this here...

File: main.tf [hunk 3]
```diff
@@ -1,18 +1,16 @@
...
 }
-
 resource "aws_ram_resource_association" "private_subnets" {
-  for_each           = toset(var.private_subnet_arns)
-  resource_arn       = each.value
+  count              = length(var.private_subnet_arns)
```
Comment (svanharmelen) [inline main.tf:8]: Are you very very sure that `count` doesn't give you the exact same error for a new deployment? As I believe it requires the same deterministic things the `for_each` requires...

**Review by svanharmelen — CHANGES_REQUESTED** [review]

File: main.tf (hunk 3 above)
Comment (costastf) [reply main.tf:8]: Yeap, tested numerous times. That is basically why the initial MR was with count and was thoroughly tested. Apparently the count gets calculated in the graph later in the graph than for_each and can be used for dynamically created resources.

File: main.tf (hunk 2 above)
Comment (costastf) [reply main.tf]: Added by accident on the MR.

File: main.tf (hunk 1 above)
Comment (costastf) [reply main.tf]: Added by accident on the MR.

File: main.tf (hunk 3 above)
Comment (dogfish182) [reply main.tf:8]: I just revalidated as well, my tests below

```
module "example_vpc_for_each" {
  source                     = "git@github.com:schubergphilis/terraform-aws-mcaf-vpc.git?ref=v1.4.0"
  name                       = "example"
  cidr_block                 = "10.112.0.0/19"
  availability_zones         = local.availability_zones
  tags                       = { environment = "example" }
  public_subnet_bits         = 22
  private_subnet_bits        = 22
  prepend_resource_type      = true
  private_s3_endpoint        = true
  share_private_subnets      = true
  share_public_subnets       = true
}

module "example_vpc_count" {
  source                     = "git@github.com:schubergphilis/terraform-aws-mcaf-vpc.git?ref=v1.4.0"
  name                       = "example"
  cidr_block                 = "10.114.0.0/19"
  availability_zones         = local.availability_zones
  tags                       = { environment = "example" }
  public_subnet_bits         = 22
  private_subnet_bits        = 22
  prepend_resource_type      = true
  private_s3_endpoint        = true
  share_private_subnets      = true
  share_public_subnets       = true
}
```

doing above and locally changing the subnet sharing module in the vpc count example produced the following

```
terraform plan -target=module.example_vpc_for_each
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

module.example_vpc_for_each.data.aws_region.current: Refreshing state...
module.example_vpc_for_each.data.aws_iam_policy_document.log_stream_action: Refreshing state...

------------------------------------------------------------------------

Warning: Resource targeting is in effect

You are creating a plan with the -target option, which means that the result
of this plan may not represent all of the changes requested by the current
configuration.

The -target option is not for routine use, and is provided only for
exceptional situations such as recovering from errors or mistakes, or when
Terraform specifically suggests to use it as part of an error message.


Error: Invalid for_each argument

  on .terraform/modules/example_vpc_for_each.subnet_sharing/main.tf line 9, in resource "aws_ram_resource_association" "private_subnets":
   9:   for_each           = toset(var.private_subnet_arns)

The "for_each" value depends on resource attributes that cannot be determined
until apply, so Terraform cannot predict how many instances will be created.
To work around this, use the -target argument to first apply only the
resources that the for_each depends on.


Error: Invalid for_each argument

  on .terraform/modules/example_vpc_for_each.subnet_sharing/main.tf line 15, in resource "aws_ram_resource_association" "public_subnets":
  15:   for_each           = toset(var.public_subnet_arns)

The "for_each" value depends on resource attributes that cannot be determined
until apply, so Terraform cannot predict how many instances will be created.
To work around this, use the -target argument to first apply only the
resources that the for_each depends on.
```

and doing it targetted against count 

```
terraform plan -target=module.example_vpc_count
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

module.example_vpc_count.data.aws_region.current: Refreshing state...
module.example_vpc_count.data.aws_iam_policy_document.log_stream_action: Refreshing state...

------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # module.example_vpc_count.aws_eip.nat[0] will be created
  + resource "aws_eip" "nat" {
      + allocation_id     = (known after apply)
+ LOTS MORE RESULTS 
```

for completeness sake I also run a full plan without targetting with the same result.

File: main.tf [hunk 4]
```diff
@@ -1,18 +1,17 @@
...
 
 resource "aws_ram_resource_association" "public_subnets" {
-  for_each           = toset(var.public_subnet_arns)
-  resource_arn       = each.value
-  resource_share_arn = aws_ram_resource_share.subnet_sharing[0].arn
-}
+  count              = length(var.public_subnet_arns)
+  resource_arn       = var.public_subnet_arns[count.index]
+  resource_share_arn = aws_ram_resource_share.subnet_sharing.arn
+}
```
Comment (svanharmelen) [inline main.tf]: Please add a newline at the end of the file...

File: outputs.tf [hunk 5]
```diff
@@ -1,9 +1,9 @@
 output "id" {
-  value       = try(aws_ram_resource_share.subnet_sharing[0].id, null)
+  value       = try(aws_ram_resource_share.subnet_sharing.id, null)
```
Comment (svanharmelen) [inline outputs.tf]: The try is not needed anymore if you do not index the resource:

```suggestion
  value       = aws_ram_resource_share.subnet_sharing.id
```

File: outputs.tf [hunk 6]
```diff
@@ -1,9 +1,9 @@
...
   description = "The ID of the subnet share in resource access manager"
 }
 
 output "arn" {
-  value       = try(aws_ram_resource_share.subnet_sharing[0].arn, null)
+  value       = try(aws_ram_resource_share.subnet_sharing.arn, null)
```
Comment (svanharmelen) [inline outputs.tf]: The try is not needed anymore if you do not index the resource:

```suggestion
  value       = aws_ram_resource_share.subnet_sharing.arn
```

File: outputs.tf [hunk 7]
```diff
@@ -1,9 +1,9 @@
...
   description = "The ARN of the subnet share in resource access manager"
 }
```
Comment (svanharmelen) [inline outputs.tf]: Please add a newline at the end of the file...

File: main.tf [hunk 8]
```diff
@@ -1,18 +1,17 @@
 resource "aws_ram_resource_share" "subnet_sharing" {
-  count                     = length(var.private_subnet_arns) > 0 || length(var.public_subnet_arns) > 0 ? 1 : 0
```
Comment (svanharmelen) [inline main.tf:2]: Are you sure this resource is _always_ needed/required? Yeah I guess it is, as otherwise why would you use this module...

**Review by svanharmelen — CHANGES_REQUESTED** [review]

File: main.tf (hunk 8 above)
Comment (dogfish182) [reply main.tf:2]: Actually that answer that you just gave is exactly why we ditched the count... it was giving us trouble and we realised that's basically all this module does, so don't use it.

File: outputs.tf (hunk 7 above)
Comment (dogfish182) [reply outputs.tf]: done

File: outputs.tf (hunk 6 above)
Comment (dogfish182) [reply outputs.tf]: done

**Review by svanharmelen — APPROVED** [review]

**Pull request merged**